---
title: "SPAN Tutorial - Shortest-Path Analysis on Networks"
author: "Max Bennett"
created: "2023-07-26"
output:
  html_document:
    df_print: paged
updated: "2023-07-26"
---


```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
defaultW <- getOption("warn") 
options(warn = -1) 
```


It's expected that you already have the packages below installed and are opening this markdown from the same directory as the helper script folder. Packages from CRAN can be downloaded using the install.packages( ) function from the console, and any other packages can be found on BioConductor.org and installed similarly from the console. This markdown was written under R version 4.2.2, Bioconductor version 3.16. 
```{r Load Dependencies, echo=TRUE, message=FALSE}

# Packages
library( circlize ) #CRAN
library( clusterProfiler ) #BioConductor
library( ComplexHeatmap ) #BioConductor
library( data.table ) #CRAN
library( doParallel ) #CRAN
library( igraph ) #CRAN
library( ggplot2 ) #CRAN
library( parallel ) #CRAN
library( pbapply ) #CRAN
library( RCy3 ) #BioConductor
library( viridis ) #CRAN

# Helper Functions  -  if you're interested to see how a function is coded you can find it in one of these files, or go directly to it using View( )
source( file.path( "Helper_R_Scripts", "shortestPathFunctions.R" ) )
source( file.path( "Helper_R_Scripts", "enrichmentTestFunctions.R" ) )
source( file.path( "Helper_R_Scripts", "ManageScriptData.R" ) )

# Enable warnings, they were suppressed while loading packages
options(warn = defaultW)
```



## Some Network Biology Basics
  **Networks** (also called graphs) are models of pairwise relationships, called **edges**, between objects called **nodes** or **vertices**. Nodes can represent abstract objects, but most of the time in biology we'll be working with nodes that directly represent genes or proteins; for the purpose of this tutorial you can think of 'node', 'gene', and 'protein' as interchangeable. Networks can have additional qualities like node or edge weights, categorical variables attached to nodes or edges, and directed edges, although none of those will be used in this tutorial. All of our network algorithms will be implemented on an un-weighted, un-directed network. 

  Many kinds of relationships between proteins can be represented with a network. The database we'll be collecting protein-protein interactions (**PPI**) from for this tutorial, STRING DB, collates information about gene fusions, gene transcript co-expression, cross-species gene co-occurrence, protein protein interactions, regulatory interactions from popular databases like KEGG, and co-occurrence of gene names in journal articles. In this tutorial we'll only be using the PPI data from STRING, which is mostly derived from high-throughput pull-down assays like yeast two-hybrid (Y2H) and affinity purification mass spectrometry (AP-MS). 
  


## Selecting the Underlying Network Data
Let's start by loading the data from STRING (there's a lot, so this might take a few minutes; if you have this data stored locally you can read it in from your local directory and skip this chunk).
```{r Load String, message=FALSE}
stringNet <- fread("https://stringdb-static.org/download/protein.links.detailed.v11.5/9606.protein.links.detailed.v11.5.txt.gz")
aliases <- fread("https://stringdb-static.org/download/protein.aliases.v11.5/9606.protein.aliases.v11.5.txt.gz")
```

Now we're going to threshold the data to only include edges with 'experimental' scores greater than 400, meaning they were found with 'medium' confidence in one or more published pulldown assays. STRING is opaque about how these confidence scores are calculated, except to say that a score approaching 1000 is extremely confident, a score of 500 expects a 50% false positive rate, and a score of 100 expects a 90% false positive rate. A 60% estimated false positive rate might not sound very good, but it's serviceable for a few reasons besides convention. High throughput PPI assays have high false positive rates approaching 30% or even 40%, so we're already not so far from our gold standard for big data PPI. While the distribution of experimental scores in STRING balloons with thresholds below 400, from 40,000 interactions with scores 400-500 to 2.5 million interactions with scores 50-150. A cutoff of 400 culls those very low confidence scores while keeping a useful number of evenly binned scores from 400-1000; the even distribution of scores over 400 suggests at least methodological consistency in those interactions. 

```{r, fig.show="hold", out.width="50%"}
hist( stringNet[experimental > 0, experimental ], br=20, col=c(0,2,2,rep(0,5),3,3,rep(0,10)) )
hist( stringNet[experimental > 400, experimental ] )
```

Ultimately, interesting linker genes will be selected based on many interactions, no single interaction is too important and the approach favors false positives over false negatives. 

igraph is a very useful package for handling networks in R, although it's data structures can be very cumbersome. After spending some time writing my own very poorly optimized code for handling networks with base R, I assimilated and built my wrapper functions using igraph for operations like finding network distance or shortest paths.
```{r Create Graph from String}
net <- stringNet[experimental >= 400] 
setnames(aliases, "#string_protein_id", "stringID")

graph <- igraph::graph_from_edgelist(as.matrix(net[, .(protein1, protein2)]))
# This calculates network distance, the number of edges in the shortest path between each pairwise combination of nodes
distances <- igraph::distances(graph)
```

Here's a visualization of the distance matrix of genes in the thresholded String network. Note: some genes are not connected at all--they'll have an infinite value in the distance matrix by default. Non-finite values aren't permitted by hclust( ) so we'll assign all Inf values the highest finite distance in the dataset. We'll keep this modification going forward.

```{r Distances visualization, message=FALSE, fig.show="hold", out.width="50%"}
distances[!is.finite(distances)] <- max(distances[is.finite(distances)])
distances_toGraph <- as.dist (distances)
ddr <- stats::hclust (distances_toGraph, method= "average") |> as.dendrogram()
hist(distances_toGraph, br=15)
ht_opt$message <- FALSE
Heatmap(distances, col = colorRamp2(c(0,2,4,6,8,15), viridis::viridis(6)), show_row_names = FALSE, show_column_names = FALSE, cluster_rows = ddr, cluster_columns = ddr, )
```



## What is Shortest-Path Analysis?
**Shortest Paths** are exactly what they sound like, the fewest nodes and edges in a network that connect 2 nodes. If the graph is weighted, the shortest path might not be the fewest nodes and edges, but the nodes and edges with the 'least cost'. Shortest paths are a useful tool in network biology for modelling signalling pathways and other sequences of cellular interactions since they can trace the interactions that connect one protein of interest to another. Signalling pathways are usually thought of starting with receptors or other signalling proteins and 'ending' with an effector protein, like a transcription factor, that drives broad changes in the cell's state---the 'start' and 'end' in signalling pathways lend themselves well to the shortest paths model. 

Of course, while a shortest path is only ever 'one protein wide,' your system of interest probably has multiple starts, multiple ends, and many proteins working in parallel to create a signalling cascade. To adapt the shortest paths model which only selects the genes between a source and a target to this paradigm, we can generate multiple shortest paths between multiple starts and ends, and take the union of those paths to create a rich subnetwork. This subnetwork includes the 'start' genes and 'end' genes and many putative linker genes that might facilitate interactions between the genes in either of those sets, but which would not be observed as significant hits in either. To select the most important linker genes in a subnetwork of shortest paths, that subnetwork can be scored on its network topography and other features to find high scoring genes.

This technique has been used in this lab to connect:  

**The interactors of disease causing proteins found by AP-MS** (such as Tuberculosis-secreted factors and oncogene mutants), which represent the likely beginning of disease pathology in the host cell  

with **CRISPR KO/KD screen results that are necessary for disease-critical phenotypes** (such as IFNB, TNAF, and iNOS levels  or  cell survival). 

In the diseased state, the disease causing proteins interact with proteins in the cell and ultimately effect some disease-related phenotype in the CRISPR-screened genes. Maybe the most critical genes are hits in both experiments, but it's much more likely and common that the interactions between these two data sets are more convoluted and modelling with shortest paths can clarify interesting results that aren't immediately obvious looking at the best hits from either experiment. You may have success using this technique for other data sets besides AP-MS and CRISPR screens.

**TLDR: This technique leverages large protein-protein interaction databases, such as STRING, to relate two data sets with a plausibly causal biological relationship.** 

```{r Load PPI and CRISPR Data}
# The PPI data is taken from Penn et al. 2018 DOI: 10.1016/j.molcel.2018.07.010
ppi <- fread(file.path("Shortest-Path_Network_Analysis_Tutorial_files", "Input_Data", "Sample_Data_PPI_TB_secretedFactors.csv") )
setnames(ppi, c("Bacterial Bait  (Locus ID)", "Human Protein Uniprot", "Human Gene name"), c("Bait", "Prey", "GeneNames")) 
ppi <- ppi[ MIST >= 0.7, ]
ppi <- ppi[, .(Prey, GeneNames), by = Bait] 
ppi[, GeneName := tstrsplit(ppi$GeneNames, " ", keep = 1)]

# The CRISPR data is taken from work by the Cox Lab at UCB and not to be used outside of this tutorial without their explicit permission
crispr <- fread(file.path("Shortest-Path_Network_Analysis_Tutorial_files", "Input_Data", "Sample_Data_CrisprKDs_TB_TNFa-IFNb.csv") )
# The crispr data is already cleaned, genes were mapped from mouse to human ("ortholog gene" is human), and thresholded to absolute fold changes > 1 and fdr-adjusted p values less than 0.05
```


```{r Map Genes to String IDs}
# Map STRING IDs to other datasets (aliases file makes it very easy to map from anything else to STRING)
crispr[ data.table( aliases[alias %in% crispr$ortholog_gene,] ), string := i.stringID, on = .(ortholog_gene = alias)]
ppi[ data.table( aliases[alias %in% ppi$Prey,] ), string := i.stringID, on = .(Prey = alias)]

# map STRING IDs to symbols 
strings2symbols <- rbind(crispr[, .(string, symbol = ortholog_gene)], ppi[, .(string, symbol = GeneName)])

message( paste0( as.character( floor( nrow(ppi[Prey %in% aliases$alias,])/nrow(ppi)*100) ), "% of genes in PPI dataset mapped to string")) 
message( paste0( as.character( floor( nrow(crispr[ortholog_gene %in% aliases$alias,])/nrow(crispr)*100) ), "% of genes in CRISPR dataset mapped to string"))
```


```{r Exclude Genes Without String Mapping}
ppi <- ppi[ string %in% rownames(distances), ]
```


## Generating the Network and Scoring It
Now that we have a basic understanding of the technique, we've prepared our underlying network, as well as our group of start nodes and end nodes, we can calculate the shortest paths and make a network from the union of those shortest paths. *find_shortestPaths()* is going to iterate through every start node, look in the distance matrix for the target node(s) that are the least distance from the start node, and then find all possible paths through the network with that distance using *igraph::shortest_paths()*. Since this is an unweighted graph, igraph will use the breadth-first search algorithm, and will usually return multiple pathways for each target with the same distance.

The *find_shortestPaths()* wrapper function returns 4 data objects
\$paths, \$nodes, \$edges, and \$startsEnds data structures:
   $paths is unlisted raw igraph paths
   $nodes is a list of all unique nodes in all paths
   $edges is a list of unique edges formatted c(a,b, b,c, c,d, e,f, etc. )
   $startsEnds is a table of the start and end node for every linker node
```{r Generate Shortest Paths}
shortestPaths <- find_ShortestPaths( startpoints = ppi$string,
                                   targets = crispr$string,
                                   graph = graph,
                                   distances = distances, 
                                   trackStartsEnds = TRUE )
```


With the shortest-paths network already figured, now we have to score it: *gen_NodeInfo( )* is a wrapper function that calculates centrality scores for each node in the network. Check out this link for an introduction to network centrality [European Bionformatics Institute - Centrality](https://www.ebi.ac.uk/training/online/courses/network-analysis-of-protein-interaction-data-an-introduction/building-and-analysing-ppins/topological-ppin-analysis/centrality-analysis/).
``` {r Calculate Centrality Measures}
nodeInfo <- gen_NodeInfo( shortestPaths )

nodeInfo [, source := "linker"] [string %in% crispr$string, source := "crispr"] [string %in% ppi$string, source := "ppi"]  [string %in% ppi$string & string %in% crispr$string, source := "both"]

nodeInfo [aliases[source == "Ensembl_EntrezGene"], symbol := i.alias, on = .(string = stringID)] 
```


To minimize the effects of underlying network topology on our measures of centrality, scores are normalized to a simulated background. The shortest-path network is simulated with the same number of starts and ends as our experimental network, but sampled randomly from the entire underlying network. This is repeated 1000 times as a good compromise between depth and time efficiency; it should take between 30 minutes and an hour to run, to just test the code without expecting accurate results, you can reduce the number of iterations to 20 or so to speed up run time. This operation is parallelized to save time; it should accomodate any operating system, but was only tested on Windows 10. Once the simulation is finished, every node's centrality score in the experimental network is divided by the mean of that node's scores from the simulation. The normalized values are thus even more abstract than the raw centrality scores, but are more reliable for selecting nodes of interest at a coarse-grained level. 
``` {r Normalize Shortest Paths and Clean Output}
pathsSimulated <- <- gen_RandomShortestPaths( nstarts = nrow( crispr ), 
                                              ntargets = nrow( ppi ), 
                                              niter = 1000, 
                                              distance = distances, 
                                              graph = graph)

nodeInfo <- normalize_ShortestPaths( pathsSimulated, nodeInfo )

```


## Visualizing and Interpereting the Results


## Some Other Considerations 
There are a few assumptions in this method that you should consider when applying it:  

- If you're using a network of PPI data, you're assuming that the interaction between your two gene sets of interest is mediated principally by complexes and allosteric interactions that are measured in pull-down assays.  

- You are assuming that your network is generally accurate for the intracellular compartment and state of cell differentiation/disease/perturbation relevant to your inquiry.

- If you implement an un-weighted network, you are using Occam's razor: that one protein's effect on another is probably mediated by the fewest interactions in a putative PPI network. Obviously, this will not always be true, so you can refine your technique as you are more selective with the edges in your underlying network, the genes admitted in either gene set, and as you apply additional features like weights to the network. 





